notebooks/Binary Classification with the UCI Credit Card Payment Default Dataset.ipynb

```diff
@@ -0,0 +1,599 @@
+   "metadata": {},
+   "source": [
+    "# The UCI Credit Card Payment Default Dataset\n",
+    "\n",
+    "The UCI datasets contains on 30,000 clients and their credit card transactions at a bank in Taiwan. In addition to static client features, the data contains the history of credit card bill payments between April and September 2005, as well as the balance limit of the client's credit card. The target is whether the client will default on a card payment in the following month, October 2005. One can imagine that a model trained on this data can be used in practice to determine whether a client is eligible for another product offering such as an auto loan. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Load the data\n",
+    "data_url = \"http://archive.ics.uci.edu/ml/machine-learning-databases/00350/default%20of%20credit%20card%20clients.xls\"\n",
+    "dataset = pd.read_excel(io=data_url, header=1).drop(columns=['ID']).rename(columns={'PAY_0':'PAY_1'})\n",
+    "dataset.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Dataset columns:\n",
+    "\n",
+    "* `LIMIT_BAL`: credit card limit, will be replaced by a biased feature\n",
+    "* `SEX, EDUCATION, MARRIAGE, AGE`: client demographic features\n",
+    "* `BILL_AMT[1-6]`: amount on bill statement for the April-September months\n",
+    "* `PAY_AMT[1-6]`: payment amount for the April-September months\n",
+    "* `default payment next month`: target, whether the customer defaulted the following month"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Sensitive feature\n",
+    "A = dataset[\"SEX\"]\n",
+    "A_str = A.map({ 2:\"female\", 1:\"male\"})\n",
+    "# Target\n",
+    "Y = dataset[\"default payment next month\"]\n",
+    "categorical_features = ['EDUCATION', 'MARRIAGE','PAY_1', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6']\n",
+    "for col in categorical_features:\n",
+    "    dataset[col] = dataset[col].astype('category')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduce Bias w.r.t. Sex\n",
+    "\n",
+    "We manipulate the balance limit feature `LIMIT_BAL`such that this feature is highly predictive for females but not for males. For example, we can imagine that a lower credit limit indicates that a female client is less likely to default, but provides no information on a male client's probability of default."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dist_scale = 0.5\n",
+    "np.random.seed(12345)\n",
+    "dataset['LIMIT_BAL'] = Y + np.random.normal(scale=dist_scale, size=dataset.shape[0])\n",
+    "dataset.loc[A==1, 'LIMIT_BAL'] = np.random.normal(scale=dist_scale, size=dataset[A==1].shape[0])"
```

Comment:
Might want a comment reminding readers that `A==1` are males, so the second line is making the feature non-predictive for them.